File: DSA/saumya/Day15_Q1.py
class Node:
  def __init__(self, data):
    self.data = data
    self.left = None
    self.right = None

def height(root):
  if root is None:
    return -1  # Base case: empty tree has height -1

  # Recursively calculate heights of left and right subtrees
  left_height = height(root.left)
  right_height = height(root.right)

  # Return the maximum height, adding 1 for the current level
  return 1 + max(left_height, right_height)

# test case
tree = Node(3)
tree.left = Node(2)
tree.left.left = Node(1)
tree.right = Node(5)
tree.right.left = Node(4)
tree.right.right = Node(6)

tree_height = height(tree)
print("Height of the tree:", tree_height) #output : 2

File: DSA/Yash_Gupta/Day_15_Q1.java
public class Day_15_Q1 {
    public static int getHeight(TreeNode root) {
        if (root == null) {return -1;}

        int left = getHeight(root.left);
        int right = getHeight(root.right);

        return Math.max(left, right) + 1;
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(2);
        root.left.left = new TreeNode(1);
        root.right = new TreeNode(5);
        root.right.left = new TreeNode(4);
        root.right.right = new TreeNode(6);

        System.out.println(getHeight(root));
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) {
        val = x;
        left = null;
        right = null;
    }
}